MTM-62669 As a User I would like to know how to use Messaging Management feature (#3163)
Cumulocity-IoT/c8y-docs: content/standard-tenant/monitoring-bundle/messaging-service.md

The features that use the Messaging Service include the microservice-based Data broker, Notifications 2.0, and the MQTT Service.

**Topics** are the core concept underlying all of the features using the Messaging Service.
A topic is a named logical channel for delivering messages from *publishers* to *subscribers*.
Each topic may have any number of publishers and subscribers, and in general, every subscriber will receive the messages sent by every publisher.
All of the subscribers on a topic will receive the published messages in the same order.
The topic will persistently store published messages until every subscriber has acknowledged that it has successfully received them.
This means that the Messaging Service can guarantee the delivery of every published message to every subscriber.

The following sections show how to monitor your tenant's usage of the Messaging Service, for each of the services that use it.

### To view the topics

Click **Messaging Service** in the **Monitoring** menu in the navigator to display a list of all features that are using the Messaging Service.

> Next to the feature name, you will also see basic information about the feature's usage of the Messaging Service, such as the number of topics, the number of publishers, and subscribers.

Reviewer: ```suggestion
Next to the feature name, you will also see basic information about the feature's usage of the Messaging Service, such as the number of topics, publishers, and subscribers.
```

Author: Done

Select a feature and click on it to see more details. This will display a list of all topics used by the feature and the limits that are applied for each of those topics.

> ![Messaging Management Topics](/images/users-guide/Administration/messaging-management-topics.png)

Reviewer: @reey, how could we automate updating these screenshots in the future?

#### Topics list

The topics list shows the following information for each topic:

> | Column name              | Description                                                                                                                                                                                                     | Alarming values | 	   	

Reviewer 2: Also a question... do we actually raise alarms when we hit these thresholds?
```suggestion
| Column name              | Description                                                                                                                                                                                                     | Alarm threshold | 	   	
```

Author: Done. Regarding alarms, we don't have them yet. I think we discussed this in one of the latest refinements, but this still has to be done...

|--------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------|

> | Name                     | Topic name, see feature-specific description of how to map it with the specific resource.                                                                                                                       | -               |

Reviewer 2: Suggestion:
```suggestion
| Name                     | Topic name. See the feature-specific documentation below for more information on mapping this to a specific source.                                                                                                                       | -               |
```

Author: Done

| Message rate out (msg/s) | Total rate of messages dispatched to the subscribers for the topic per second. Dispatch includes additional batching and queuing mechanisms, so this rate could differ from the subscriber acknowledgment rate. | -               |
| Subscribers              | Total number of registered subscribers. This includes both actively consuming subscribers and those that are currently disconnected and not consuming any messages.                                             | > 5             |
| Message backlog          | Backlog size in bytes which corresponds to the size occupied by unconsumed messages.                                                                                                                            | > 20 MB         |
| Used backlog             | Percentage usage of the backlog quota limit.                                                                                                                                                                    | > 80%           |

> Please refer to the feature-specific documentation for more information on how to map the topic with the source and how to clear the backlog when reaching the alarming values.

Reviewer 2: Suggestion:
```suggestion
Please refer to the feature-specific documentation below for more information on how to map the topic name to the source, and how to clear the backlog when reaching the alarm threshold.
```

Author: Done

#### Messaging Service limits

> All the backlog limits that are visible at the top of the topics list view are applied per topic, meaning if the backlog quota is set to 25MB, each topic will queue the messages until they reach the configured limit.

Reviewer: Who can configure it and how? Can a tenant administrator do this?

Author: Added info: `Limits are {{< product-c8y-iot >}} platform-wide, and only the Operations team can change them.`

### To view the topic details

Click on a selected topic name to navigate to the topic details view.
The view contains information about the topic at the top and the list of all subscribers for that topic below.

![Messaging Management Topic Details](/images/users-guide/Administration/messaging-management-topic-details.png)

#### Subscribers list

The subscribers list shows the following information for each subscriber:

| Column name                 | Description                                                                                                   | Alarming values | 	   	
|-----------------------------|---------------------------------------------------------------------------------------------------------------|-----------------|

> | Name                        | Subscriber name, see feature-specific description of how to map it with the specific resource.                | -               |

Reviewer 2: Suggestion:
```suggestion
| Column name                 | Description                                                                                                   | Alarm threshold | 	   	
|-----------------------------|---------------------------------------------------------------------------------------------------------------|-----------------|
| Name                        | Subscriber name. See the feature-specific documentation below for more information on mapping this to a specific destination.                | -               |
```

Author: Done

| Connected clients           | Number of clients that are currently connected and consuming messages.                                        | -               |
| Acknowledgment rate (msg/s) | Current rate per second of messages fully processed (consumed, processed, and acknowledged) by the consumers. | -               |
| Last acknowledged           | Latest timestamp when the message was fully processed by the consumer.                                        | >= 1 day        |
| Unacknowledged messages     | Number of unconsumed messages.                                                                                | > 1000          |
| Used backlog                | Percentage usage of the backlog quota limit by the subscriber.                                                | > 80%           |

> Please refer to the feature-specific documentation for more information on how to map the subscriber name with the source and how to clear the backlog when reaching the alarming values.

Reviewer 2: Suggestion:
```suggestion
Please refer to the feature-specific documentation below for more information on how to map the subscriber name to the destination and how to clear the backlog when reaching the alarm threshold.
```

Author: Done

### Monitoring Notifications 2.0 {#monitoring-notifications-2.0}

#### Topic and subscriber

> The topic name is the same as the `subscription` field used in the [Notifications 2.0 Subscriptions API](https://{{< domain-c8y >}}/api/core/#operation/postNotificationSubscriptionResource).

Reviewer 2: Suggestion:
```suggestion
The topic name is the same as the `subscription` field used in the [Notifications 2.0 Subscriptions API](https://{{< domain-c8y >}}/api/core/#operation/postNotificationSubscriptionResource) and the [Notifications 2.0 Tokens API](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenResource).
```

Author: Done

The subscriber name is the same as the `subscriber` field used in the [Notifications 2.0 Tokens API](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenResource).

Both the topic and the subscriber are created only when there is a web socket connection established to the Notifications 2.0.

> After that, the Messaging Service will collect the messages under the given topic until they are consumed, they reach the configured time-to-live interval, or the [subscriber is unsubscribed](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenUnsubscribeResource) for the topic.

Reviewer 2: Suggestion (I think this is correct):
```suggestion
The subscriber is created the first time that a Notifications 2.0 WebSocket connection is established using a token with given subscription and subscriber names.
The topic itself is created the first time that _any_ Notifications 2.0 WebSocket connection is established using a token with the given subscription name.
However, once the subscriber is created it will not be deleted even if the WebSocket connection is disconnected.
That is, the Messaging Service will collect and persist the messages under the given topic until either they are consumed, they reach the configured time-to-live (TTL) interval, or the [subscriber is explicitly unsubscribed](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenUnsubscribeResource) from the topic.
```

Author: Yes, everything is correct. Suggestion applied.

Please refer to the [Consumer lifecycle](https://{{< domain-c8y >}}/api/core/#section/Overview/Consumer-lifecycle) for more details.

> #### Clearing the backlog

Reviewer 2: Praise: This part (and the same for the MQTT Service) is really nice - very clearly explained.

There are a few ways to clear the backlog from the Notifications 2.0 topics.

##### Consume messages

If the topic and subscriber were created, there are probably also valuable messages that are stored in the Messaging Service and should be consumed.
To consume and acknowledge the messages for a given topic and subscriber:
* Create the [Notifications 2.0 Token](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenResource) for the selected topic and subscriber.

> * Use the token to create the [web socket connection to the Notifications 2.0](https://{{< domain-c8y >}}/api/core/#section/Consumer-protocol).

Reviewer 2: Suggestion:
```suggestion
* Use the token to create a [Notifications 2.0m WebSocket connection](https://{{< domain-c8y >}}/api/core/#section/Consumer-protocol) to the topic.
```

Author: Done

This will remove the messages from the Messaging Service and clear the backlog for the given topic and subscriber, but the action is not permanent.
As the Notifications 2.0 subscription and the subscriber still exist, the backlog can be filled again with new messages if they are not consumed continuously.

##### Unsubscribe the subscriber using Notifications 2.0 API {#unsubscribe-the-subscriber-using-notifications-2.0-api}

If the subscriber is not needed anymore and there are no valuable messages that should be consumed, the subscriber can be unsubscribed.
To do this:
* Create the [Notifications 2.0 Token](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenResource) for the selected topic and subscriber.
* Use the token to unsubscribe the subscriber from the topic by calling the [Notifications 2.0 Token Unsubscribe API](https://{{< domain-c8y >}}/api/core/#operation/postNotificationTokenUnsubscribeResource).

This will remove the subscriber from the Messaging Service and clear the backlog for the given subscriber and the whole topic if there are no more subscribers with unconsumed messages.

> If the subscriber won't be recreated by establishing the [web socket connection to the Notifications 2.0](https://{{< domain-c8y >}}/api/core/#section/Consumer-protocol), this action is permanent, meaning the backlog won't grow again.

Reviewer 2: Suggestion:
```suggestion
If the subscriber is not recreated by establishing a [Notifications 2.0 WebSocket connection](https://{{< domain-c8y >}}/api/core/#section/Consumer-protocol) to the topic, this action is permanent, meaning the backlog won't grow again.
```

Author: Done

If there are no more active subscribers, it is also recommended to delete the [Notifications 2.0 Subscription](https://{{< domain-c8y >}}/api/core/#operation/deleteNotificationSubscriptionResource).

##### Unsubscribe the subscriber from the Messaging Management View

If the subscriber is not needed anymore and there are no valuable messages that should be consumed, the subscriber can be unsubscribed.
To do this, select the subscriber from the subscriber list in the Messaging Service view and click the unsubscribe icon.

> This action is equivalent to the [Unsubscribe the subscriber using Notifications 2.0 API](#unsubscribe-the-subscriber-using-notifications-2.0-api).

Reviewer 2: Suggestion:
```suggestion
This action is equivalent to [unsubscribing the subscriber using Notifications 2.0 API](#unsubscribe-the-subscriber-using-notifications-2.0-api).
```

Author: Done

All the information about permanent action and cleaning are the same.

### Monitoring the MQTT Service {#monitoring-the-mqtt-service}

#### Topic and subscriber

> The topic name is mapped 1:1 with the topic from the MQTT Service.

Reviewer 2: Suggestion:
```suggestion
The topic name is mapped 1:1 to the topic name used by the MQTT Service client.
```

Author: Done

After consuming all the messages, the backlog will be cleared and the topic will be ready to store new messages.

##### Unsubscribe the subscriber using MQTT Service SDK {#unsubscribe-the-subscriber-using-mqtt-service-sdk}

If the subscriber is not needed anymore and there are no valuable messages that should be consumed, the subscriber can be unsubscribed.
Use the [unsubscribe action]({{< link-c8y-github >}}/cumulocity-clients-java/blob/develop/mqtt-service/websocket/src/main/java/com/cumulocity/mqtt/service/sdk/websocket/WebSocketSubscriber.java#L60) from the MQTT Service SDK.
This will remove the subscriber from the Messaging Service and clear the backlog for the given subscriber and the whole topic if there are no more subscribers with unconsumed messages.

##### Unsubscribe the subscriber from the Messaging Management View

If the subscriber is not needed anymore and there are no valuable messages that should be consumed, the subscriber can be unsubscribed.
To do this, select the subscriber from the subscriber list in the Messaging Service view and click the unsubscribe icon.

> This action is equivalent to the [Unsubscribe the subscriber using MQTT Service SDK](#unsubscribe-the-subscriber-using-mqtt-service-sdk).

Reviewer 2: Suggestion:
```suggestion
This action is equivalent to [unsubscribing the subscriber using MQTT Service SDK](#unsubscribe-the-subscriber-using-mqtt-service-sdk).
```

Author: Done
